core/shared/vector_list.inc

```diff
@@ -0,0 +1,56 @@
+/**************************************************************************/
+/*  vector_list.inc                                                       */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef SHARED_VECTOR_LIST
+
+#if defined(VECTOR_H) && defined(LIST_H)
+
+#define SHARED_VECTOR_LIST
+
+template <typename T>
+template <typename A>
+Vector<T>::Vector(const List<T, A> &p_list) {
```

Comment:
I get _why_ this works, but why is this needed over just forward declarations (which you also do use)?

Reply:
Because the logic for fully-generic, templated functions have to be declared in a header *somewhere*. This is the only way (to my knowledge) to define forward declared logic between two separate files *without* requiring an entirely separate header include.

A more practical use-case would be applying this logic to `constexpr` structs, as that would allow them to fully interop in `constexpr` contexts. So something like a sanity check, `static_assert(Vector2(0, 0) == Vector2i(0, 0))`, could be included in that file outright to make sure everything is properly synced

Reply:
Ah, so it does! So this technique actually isn't necessary with generic templates. I was applying it out of habit because this *is* required for fully-declared types; so `Vector2`/`Vector2i` would use this if they became `constexpr`.

That thankfully simplifies matters greatly, but means my initial writeup isn't necessary anymore, oops. Gonna reuse that for some `constexpr` math PR I'll inevitably cook up in the future.